src/transactions/TransactionFrame.cpp

```diff
@@ -1750,8 +1755,11 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
             }
             ++opNum;
 
-            bool txRes = op->apply(app, signatureChecker, ltxOp, subSeed,
-                                   opResult, txResult.getSorobanData());
+            evtManager.clear();
```

Comment:
Is this `clear` necessary? It seems like semantically it would make more sense to `clear` on failure after emitting diagnostics, and asserting that the events manager is empty here right?

Reusing the `EventsManager` like this is also more error-prone that an object specifically for this operation. Did you consider passing a different type to the operations to collect the events?

Reply:
Yeah, clearing the whole data structure seems sketchy. I like the idea of using a separate struct, especially given that we're going to have two distinct event locations (tx events may be only fee events, remaining events are operation events). The helper should also probably hold a reference to the parent tx event manager and use it for recording the diagnostic events, so that you don't need to flush them every time.

Reply:
I've reworked the design to incorporate the separation between tx- and op- level event management.